### PR TITLE
Remove global var Handler from virt-launcher/virtwrap/network

### PR DESCRIPTION
**Pull request by qinqon**
**What this PR does / why we need it**:
The virt-launcher network code contains a pair of global elements to
open the door for mocking, a `Handler` variable and a `podNICFactory`,
using global variables is in general bad practice. This change replace
those with structs constructed at the caller making code easier to
follow and to test.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This is part of the CleanCode initiative.
Depends on https://github.com/kubevirt/kubevirt/pull/5512

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/network/common.go [hunk 1]
```diff
@@ -528,10 +526,8 @@ func (h *NetworkUtilsHandler) DisableTXOffloadChecksum(ifaceName string) error {
 var DHCPServer = dhcp.SingleClientDHCPServer
 var DHCPv6Server = dhcpv6.SingleClientDHCPv6Server
 
-func initHandler() {
-	if Handler == nil {
-		Handler = &NetworkUtilsHandler{}
-	}
+func NewHandler() NetworkHandler {
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/common.go:529]: [Returning an interface instead of the concrete](https://ehaas.net/blog/the-go-interface/#return-an-interface) is usually not the right thing to do.
I would suggest returning the concrete struct and let the caller assign it to an interface variable.

File: pkg/virt-launcher/virtwrap/network/network.go [hunk 2]
```diff
@@ -66,6 +65,14 @@ type podNIC interface {
 	PlugPhase2(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, domain *api.Domain, podInterfaceName string) error
 }
 
+type Network struct {
+	podnic podNIC
+}
+
+func New() *Network {
+	return &Network{podnic: &podNICImpl{networkHandler: NewHandler()}}
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/network.go:73]: I guess this is the core change in this PR.

You structured a Network that has a podNIC member and the podNIC having a Handler.
I'm having hard time to compute this.

Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:73]: BTW, why do we need to instantiate a `Network`? I do not see it used in any interesting manner.

Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go:73]: More or less we have three levels of expose:
- Network struct: expose the functions to the package
- podNIC: expose the Phase1/Phase2 internal functions 
- networkHandler: expose the low level network functions so we can mock them.

**Comment by qinqon** [comment]
[force-push](https://github.com/kubevirt/kubevirt/compare/a114e1128253f58a647e2566d0ce75364d64fd0b..8ba5a3c9b4edc8d0643590c1e0277d7abaf6c696): Use concrete type for NetworkHandler, create `findNetwork` helper

File: pkg/virt-launcher/virtwrap/network/network.go (hunk 2 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go:73]: > BTW, why do we need to instantiate a Network? I do not see it used in any interesting manner.

Is the trick to test the `SetupPodNetwork` functions

```
		podNICFactory = newpodNIC
		networkMock = &Network{podnic: mockpodNIC}
	})

	Context("interface configuration", func() {
		It("should configure bridged pod networking by default", func() {
			podNICFactory = func(network *v1.Network) (podNIC, error) {
				return mockpodNIC, nil
			}
			vm := newVMIBridgeInterface("testnamespace", "testVmName")
			iface := v1.DefaultBridgeNetworkInterface()
			defaultNet := v1.DefaultPodNetwork()

			mockpodNIC.EXPECT().PlugPhase1(vm, iface, defaultNet, primaryPodInterfaceName, pid)
			err := SetupPodNetworkPhase1(vm, pid)
			err := networkMock.SetupPodNetworkPhase1(vm, pid)
			Expect(err).To(BeNil())
		})

```

Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:73]: I'm very happy with this PR; I had it as a rough idea and thought of it as a follow up to my PR #4990, but it considered it too difficult for me.

However, I'm not sure about the name and merit of the suggested `Network` type. It has nothing to do with `v1.Network` or any other specific "network" in the code. Frankly, I don't see what it adds on top of the `podNIC` type. If I understand it correctly, it exists only to serve as a receiver for `SetupPodNetworkPhase1`. Would you consider shoving `podNIC` as the receiver, instead? `podNIC` is a badly-named type - but that's a good thing, because it would not confuse the reader to think that it is something else.

Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:73]: I think that the `NetworkUtilsHandler` can also be embedded into the `podNICImpl` so the operations will be executed on this strange object and not have it as member.

Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go:73]: I will try to embedded structs a  little to see if things get simpler also, `NetworkHandler` can be named `NetworkCommands` and we can have the `LinuxNetworkCommands` implementation (at the end is what we have more or less)

**Comment by qinqon** [comment]
@EdDev @dankenigsberg I have reworked the PR so it "only" remove the Handler global but not the factory so PR is smaller and easier to understand, I will do a follow up on the factory.

**Comment by dankenigsberg** [comment]
I hate github; I had a comment today about finding `driver: Handler` confusing. My comment vanished. I think that the mental burden of `handler: Handler` is smaller.

**Comment by qinqon** [comment]
[force-push](https://github.com/kubevirt/kubevirt/compare/be04dc3cd9e039d5858adc8ae0377f8975035506..af04735d4de19b05f544e461fac3df0098b0b44f): renaming "driver" with "handler"

File: pkg/virt-launcher/virtwrap/network/network.go [hunk 3]
```diff
@@ -131,7 +131,7 @@ func SetupPodNetworkPhase2(vmi *v1.VirtualMachineInstance, domain *api.Domain) e
```
Comment (dankenigsberg) [inline pkg/virt-launcher/virtwrap/network/network.go:131]: I don't understand the title and message of this commit; and neither it's motivation

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: pkg/virt-launcher/virtwrap/network/network.go (hunk 3 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go:131]: What's the name of it ?

Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go:131]: Added description

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 4]
```diff
@@ -69,7 +69,9 @@ type BindMechanism interface {
 	startDHCP(vmi *v1.VirtualMachineInstance) error
 }
 
-type podNICImpl struct{}
+type podNICImpl struct {
+	driver NetworkHandler
```
Comment (dankenigsberg) [inline pkg/virt-launcher/virtwrap/network/podinterface.go]: but why is it still a `driver` rather a `handler`, @qinqon. I don't know if it is a github quirk that hides my question from you :-(.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 4 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: This is old, is no longer handler, please review latest version.

File: pkg/virt-launcher/virtwrap/network/common.go [hunk 5]
```diff
@@ -196,11 +196,11 @@ func (h *NetworkUtilsHandler) ConfigureIpForwarding(proto iptables.Protocol) err
 }
 
 func (h *NetworkUtilsHandler) IsIpv6Enabled(interfaceName string) (bool, error) {
-	link, err := Handler.LinkByName(interfaceName)
+	link, err := h.LinkByName(interfaceName)
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/network/common.go:197]: The commit message confused me; could you indicate you're using the receiver ? The 'itself' word threw me off.

File: pkg/virt-launcher/virtwrap/network/network.go (hunk 3 above)
Comment (maiqueb) [reply pkg/virt-launcher/virtwrap/network/network.go:131]: Yeah, the commit msg is weird. Promote ? 

I would instead specify that the network handler is now a member of the podNic struct.

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 6]
```diff
@@ -83,17 +85,17 @@ func writeVifFile(buf []byte, pid, name string) error {
 	return nil
 }
 
-func setPodInterfaceCache(iface *v1.Interface, podInterfaceName string, uid string) error {
+func (l *podNICImpl) setPodInterfaceCache(iface *v1.Interface, podInterfaceName string, uid string) error {
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/network/podinterface.go:88]: If you're moving these functions to be methods of the `podNICImpl` struct, shouldn't you also set iface / podInterfaceName / UID as members of the struct ? 

That way, you can clean the method's signature.

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 7]
```diff
@@ -70,7 +70,7 @@ type BindMechanism interface {
 }
 
 type podNICImpl struct {
-	driver NetworkHandler
+	handler NetworkHandler
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/network/podinterface.go:73]: I think this change should be done in the previous commit. That would address @dankenigsberg 's comment in https://github.com/kubevirt/kubevirt/pull/4977/commits/d7fbf2cba4ce4ede1f1ccdcd4df30eb726c4e40d#r580912130 .

**Review by maiqueb — CHANGES_REQUESTED** [review]
Thanks, very decent improvement on this murky part of the code.

I still request you to explain a little bit what you're doing in the commit msgs (while the code reads very well, the commit msgs are a bit confusing).

Furthermore, I wonder if you should take the opportunity to make the arguments of the  `podNicImpl` methods as members of the struct instead.

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 4 above)
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: oh, I think it's because a wrong split to commits. You introduce `driver` in one, and change it to `handler` in another - which confused me.

Beside this, I think this PR is ok (but I'm not sure I'm qualified to really approve this grand change)

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 7 above)
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:73]: you've beat me to it!

File: pkg/virt-launcher/virtwrap/network/common.go (hunk 5 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/common.go:197]: Done

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 6 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:88]: I am doing that at next PR that remove the pidfactory alltogether https://github.com/kubevirt/kubevirt/pull/5062/commits/dd68d660535f9802bd2bcb9ef2a283d8ab2fa7fe#diff-030aa8b7ff9bf4a511d1d5641c1dec04205a944137672be3a691f0b1b4068753R87-R92

File: pkg/virt-launcher/virtwrap/network/network.go (hunk 3 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go:131]: Done

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 4 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: Done

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 7 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:73]: Done, removed driver from whole PR.

**Comment by qinqon** [comment]
[force-push](https://github.com/kubevirt/kubevirt/compare/718e658a44f9ea759a9667995c325973e7501269..7a6002ddd977a2dbd6764a7cd5ad4159c3893fad): Addressed @maiqueb and @dankenigsberg comments

**Review by maiqueb — APPROVED** [review]
Thanks.

**Comment by qinqon** [comment]
@maiqueb rebased on master.

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go [hunk 8]
```diff
@@ -287,15 +292,15 @@ var _ = Describe("Pod Network", func() {
 		Expect(err).To(BeNil())
 	}
 
-	TestRunPlug := func(driver BindMechanism) {
-		err := driver.discoverPodNetworkInterface()
+	TestRunPlug := func(bindMechanism BindMechanism) {
+		err := bindMechanism.discoverPodNetworkInterface()
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/network/podinterface_test.go:296]: This rename has nothing to do with this change, does it ? Your commit message is:

"Making the handler a member of the podNICImpl struct enables us to properly
mock it in hte unit tests"

And looking below looks like most of the changes in this commit originate from this rename. 

I think you should either leave renaming out of the commit, or update the commit msg.

**Review by maiqueb — CHANGES_REQUESTED** [review]

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 8 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go]: Going back to "driver" to make the PR smaller, the renaming can be done at a follow up if needed

**Review by maiqueb — APPROVED** [review]
Thanks

**Comment by qinqon** [comment]
@maiqueb rebased can you take a look again ?

**Review by maiqueb — APPROVED** [review]

**Review by EdDev — COMMENTED** [review]
commit: refactor: User receiver at NetworkHandlerUtils 

- Could you set the topic to: `virt-launcher, network: `?
- You named in the text `NetworkHandlerUtils`, but it is `NetworkUtilsHandler`.

**Review by EdDev — COMMENTED** [review]
commit: refactor: Move handler as podNICImpl member

typo: `hte`

Also, this commit does more than placing the handler under `podNICImpl`, it places some other functions under it.

**Comment by qinqon** [comment]
Rebased and fixed @EdDev comments on commist messages.

**Review by maiqueb — APPROVED** [review]

**Comment by qinqon** [comment]
[force-push](https://github.com/kubevirt/kubevirt/compare/1fbb9f0fe667db28285a584793909e41604cce31..4564be996bff44f72592fdd95faf04174bb7c345): Rebased

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go [hunk 9]
```diff
@@ -92,6 +93,10 @@ var _ = Describe("Pod Network", func() {
 		ctrl = gomock.NewController(GinkgoT())
 		mockNetwork = NewMockNetworkHandler(ctrl)
 		Handler = mockNetwork
+		podnic = podNICImpl{cacheFactory: cacheFactory, handler: mockNetwork}
+		podNICFactory = func(cacheFactory cache.InterfaceCacheFactory) podNIC {
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/podinterface_test.go:96]: Why is this needed?

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 10]
```diff
@@ -71,6 +71,7 @@ type BindMechanism interface {
 
 type podNICImpl struct {
 	cacheFactory cache.InterfaceCacheFactory
+	handler      NetworkHandler
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/podinterface.go:74]: Can you please explain what do we gain by having this change?
You wrote -
>> enables us to properly mock it in the unit tests
I couldn't see any special change in the unit tests.

If we want to get rid of the `Handler` global, I would change the `Handler` to be a singleton.

File: pkg/virt-launcher/virtwrap/network/common.go [hunk 11]
```diff
@@ -194,11 +194,11 @@ func (h *NetworkUtilsHandler) ConfigureIpForwarding(proto iptables.Protocol) err
...
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/common.go:195]: Since we have a discussion about the other commits in this PR. This one is for sure correct and needed, it can be moved to another PR to be merged faster.

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 12]
```diff
@@ -123,15 +124,15 @@ func setPodInterfaceCache(iface *v1.Interface, podInterfaceName string, vmi *v1.
 	return nil
 }
 
-func readIPAddressesFromLink(podInterfaceName string) (string, string, error) {
-	link, err := Handler.LinkByName(podInterfaceName)
+func (l *podNICImpl) readIPAddressesFromLink(podInterfaceName string) (string, string, error) {
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/podinterface.go:127]: I don't see any connection between this method and `podNICImpl`. It is just a helper method for some of the bindings.
I find it very strange that `podNICImpl` created the `BindingMechanism` and also has method used by the `BindingMechanism` it created.

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 9 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go:96]: This is used here https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/network/network.go#L38 and is the kind of bad pattern I want to remove (podNICFactory is removed at different PR though). 

In general globals or singletons are something to try to avoid, they hide dependencies like this, making difficult to pass a fake instance of them to the code you want to test, also the hide the fact that a global instance is being used so in case it has some state this state is going to be shared between instance using the singleton.

From Brian Button:

```
1) Singletons frequently are used to provide a global access point for some service.
True, they do this, but at what cost? They provide a well-known point of access to some service in your application so that you don't have to pass around a reference to that service. How is that different from a global variable? (remember, globals are bad, right???) What ends up happening is that the dependencies in your design are hidden inside the code, and not visible by examining the interfaces of your classes and methods. You have to inspect the code to understand exactly what other objects your class uses. This is less clear than it could be. The urge to create something as a global to avoid passing it around is a smell in your design; it is not a feature of globals/singletons. If you examine your design more closely, you can almost always come up with a design that it is better and does not have to pass around tramp data to every object and method.

2) Singletons allow you to limit creation of your objects.
This is true, but now you are mixing two different responsibilities into the same class, which is a violation of the Single Responsibility Principle. A class should not care whether or not it is a singleton. It should be concerned with its business responsibilities only. If you want to limit the ability to instantiate some class, create a factory or builder object that encapsulates creation, and in there, limit creation as you wish. Now the responsibilities of creation are partitioned away from the responsibilities of the business entity.

3) Singletons promote tight coupling between classes
One of the underlying properties that makes code testable is that it is loosely coupled to its surroundings. This property allows you to substitute alternate implementations for collaborators during testing to achieve specific testing goals (think mock objects). Singletons tightly couple you to the exact type of the singleton object, removing the opportunity to use polymorphism to substitute an alternative. A better alternative, as discussed in the first point above, is to alter your design to allow you to pass references to objects to your classes and methods, which will reduce the coupling issues described above.

4) Singletons carry state with them that last as long as the program lasts
Persistent state is the enemy of unit testing. One of the things that makes unit testing effective is that each test has to be independent of all the others. If this is not true, then the order in which the tests run affects the outcome of the tests. This can lead to cases where tests fail when they shouldn't, and even worse, it can lead to tests that pass just because of the order in which they were run. This can hide bugs and is evil. Avoiding static variables is a good way to prevent state from being carried from test to test. Singletons, by their very nature, depend on an instance that is held in a static variable. This is an invitation for test-dependence. Avoid this by passing references to objects to your classes and methods.
```

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 10 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:74]: > Can you please explain what do we gain by having this change?
> You wrote -
> 
> > enables us to properly mock it in the unit tests
> 
> I couldn't see any special change in the unit tests.
Now that handler is a clear dependency of podnic we just pass a mock to the dependency so we don't have to change a global or rewrite a singleton initialize functioon or the like.

```golang
mockNetwork = NewMockNetworkHandler(ctrl)                               
        podnic = podNICImpl{cacheFactory: cacheFactory, handler: mockNetwork}   
```

> 
> If we want to get rid of the `Handler` global, I would change the `Handler` to be a singleton.

ditto https://github.com/kubevirt/kubevirt/pull/4977#discussion_r618289722

File: pkg/virt-launcher/virtwrap/network/common.go (hunk 11 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/common.go:195]: Good point will do that.

Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/common.go:195]: Done https://github.com/kubevirt/kubevirt/pull/5512

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 9 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go:96]: I still don't understand why we need to override `podNICFactory` here.

**Comment by qinqon** [comment]
Rebased in top of https://github.com/kubevirt/kubevirt/pull/5512

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 9 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go:96]: Ok, I think I got it. The issue is that we are invoking `SetupPodNetworkPhase1` that is `network.go` method from the unit tests of `podinterface_test.go`. We should invoke `PlugPhase1`, it is a unit test, we shouldn't test `network.go` here.

https://github.com/kubevirt/kubevirt/blob/6b367b3804d2d045528344d6cfac8a4f5cd3e21d/pkg/virt-launcher/virtwrap
/network/podinterface_test.go#L284
https://github.com/kubevirt/kubevirt/blob/6b367b3804d2d045528344d6cfac8a4f5cd3e21d/pkg/virt-launcher/virtwrap/network/podinterface_test.go#L343

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 10 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:74]: `common.go` is a utility class with helper methods. The only reason we have the `NetworkHandler` is because of our tests limitations to mock global methods.

I find it very confusing to pass a struct that has no state and is there just because of a tests limitation as a parameter to so many methods. It gives the feeling that the may be several implementations of the `handler`. I find it very confusing.

So yes, globals most of the times are a bad practice. But in this case, IMO, the proposed alternative is not better.

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 9 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go]: Done at different commit https://github.com/kubevirt/kubevirt/pull/4977/commits/9535342fc593f8a117a19a702b07cd156d59c641.

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 10 above)
Comment (maiqueb) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:74]: @AlonaKaplan , my 2 cents:

> > `common.go` is a utility class with helper methods. The only reason we have the `NetworkHandler` is because of our tests limitations to mock global methods.
> 
> also it decouples low level primitives, for example if we want to use different library than "netlink" it will help with that.
> 
> > I find it very confusing to pass a struct that has no state and is there just because of a tests limitation as a parameter to so many methods.

This struct already exists. I think @qinqon is just making it explicit, instead of keeping it hidden behind smoke and mirrors.

> 
> Only `createAndBindTapToBridge` receive the handler at the function, the rest use it from the podnic or from the binding struct.
> 
> ```
> podinterface.go:        err := createAndBindTapToBridge(b.handler, tapDeviceName, b.bridgeInterfaceName, b.queueCount, *b.launcherPID, int(b.vif.Mtu), "0")
> ```

Not sure this would help, but you could take this refactor a step further and add the `createAndBindTapToBridge` to the bidge / masquerade struct implementations and make it a method - that way, you'd not need to inject the network driver into it.

> 
> > It gives the feeling that the may be several implementations of the `handler`. I find it very confusing.

The feeling that there might be several implementations of the handler should come from the fact that the handler is an interface. Which it already is - and has been from the start.

> 
> There has being always two implementations the real one and the mock one, and hiding this with a global var make thing even more confusing, if we think about the issue we have at this PR trying to find why podNICFactory was needed podinterface_test.go it illustrate the problem I think.
> 
> > So yes, globals most of the times are a bad practice. But in this case, IMO, the proposed alternative is not better.
> 
> I understand that looks a little too much to promote handler as podnic/bindmechanism member just for unit test, but in fact what we are doing here is to make clear the separation we were hidding with global vars, and IMO that's always a good thing.

+1

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go [hunk 13]
```diff
@@ -93,9 +93,6 @@ var _ = Describe("Pod Network", func() {
 		ctrl = gomock.NewController(GinkgoT())
 		mockNetwork = NewMockNetworkHandler(ctrl)
 		podnic = podNICImpl{cacheFactory: cacheFactory, handler: mockNetwork}
-		podNICFactory = func(cacheFactory cache.InterfaceCacheFactory) podNIC {
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/podinterface_test.go]: It should be removed from the `AfterEach` as well.

**Comment by qinqon** [comment]
rebased

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 13 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go]: Done

**Comment by qinqon** [comment]
[force-push](https://github.com/kubevirt/kubevirt/compare/83368c97ace8b6bc2333e91fed9adb7019290441..cd99207e743fd8fd67fd71e60147a20ac0a6daff): Addressed @AlonaKaplan comments

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go [hunk 14]
```diff
@@ -93,7 +93,6 @@ var _ = Describe("Pod Network", func() {
 
 		ctrl = gomock.NewController(GinkgoT())
 		mockNetwork = NewMockNetworkHandler(ctrl)
-		Handler = mockNetwork
```
Comment (maiqueb) [inline pkg/virt-launcher/virtwrap/network/podinterface_test.go:95]: shouldn't this be part of the next commit  ?

**Review by maiqueb — CHANGES_REQUESTED** [review]
I'm having some trouble understanding why the 3rd commit touches the `podinterface_test` file altogether. 

Can't you squash those changes into the last commit ?

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 14 above)
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go:95]: This commit remove the "Handler" variable so we have to remove it from the test too

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go [hunk 15]
```diff
@@ -164,6 +163,7 @@ var _ = Describe("Pod Network", func() {
 	})
 
 	AfterEach(func() {
+		podNICFactory = newpodNIC
```
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go]: This fixes a bug for unit test, they were not setting the factory back

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 14 above)
Comment (maiqueb) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go:95]: Of course. Thanks.

File: pkg/virt-launcher/virtwrap/network/podinterface_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/podinterface_test.go]: fair enough, removed this line from the affected commit.

**Review by maiqueb — APPROVED** [review]
Thanks.

File: pkg/virt-launcher/virtwrap/network/network.go [hunk 16]
```diff
@@ -144,5 +144,5 @@ func isSecondaryMultusNetwork(net v1.Network) bool {
 }
 
 func newpodNIC(cacheFactory cache.InterfaceCacheFactory) podNIC {
-	return &podNICImpl{cacheFactory: cacheFactory, handler: Handler}
+	return &podNICImpl{cacheFactory: cacheFactory, handler: &NetworkUtilsHandler{}}
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/network.go:147]: I find it weird that a vmi/pod with multiple nics will have multiple instances of `NetworkUtilsHandler`.

Comment (qinqon) [reply pkg/virt-launcher/virtwrap/network/network.go]: I have prome it as podnic factory method arg, now the instance is reused between pod nics.

**Comment by qinqon** [comment]
rebased and addressed @AlonaKaplan comment

**Review by maiqueb — CHANGES_REQUESTED** [review]
I do not like (borderline don't understand) the term "promore" on commit msg to the last commit.

I had said so in https://github.com/kubevirt/kubevirt/pull/4977#discussion_r582781613 , on a different attribute.

Anyway, it at least has a typo: "Promete handler as pod factory method arg".

**Review by maiqueb — APPROVED** [review]
Thank you.

**Comment by qinqon** [comment]
/retest
hiccups
